src/hypergraph.jl

```diff
@@ -14,35 +14,61 @@ A hypergraph storing information about vertices and hyperedges.
 **Constructors**
 
     Hypergraph{T}(n,k) where {T<:Real}
+    Hypergraph{T,V}(n, k) where {T<:Real, V}
+    Hypergraph{T,V,E}(n, k) where {T<:Real, V, E}
 
 Construct a hypergraph with a given number of vertices and hyperedges.
+Optionally, values of type `V` can be stored at vertices and values of type `E`
+can be stored at hyperedges.
 
-    Hypergraph(m::AbstractMatrix{T}) where {T<:Real}
+    Hypergraph(m::AbstractMatrix{Union{T, Nothing}}) where {T<:Real}
+    Hypergraph{V}(m::AbstractMatrix{Union{T, Nothing}}) where {T<:Real, V}
+    Hypergraph{V, E}(m::AbstractMatrix{Union{T, Nothing}}) where {T<:Real, V, E}
 
 Construct a hypergraph using its matrix representation.
 In the matrix representation rows are vertices and columns are hyperedges.
+Optionally, values of type `V` can be stored at vertices and values of type `E`
+can be stored at hyperedges.
 
 **Arguments**
 
-* `T` : type of values stored in the hypergraph
+* `T` : type of weight values stored in the hypergraph
+* `V` : type of values stored in the vertices of the hypergraph
+* `E` : type of values stored in the edges of the hypergraph
 * `n` : number of vertices
 * `k` : number of hyperedges
 * `m` : a matrix representation rows are vertices and columns are hyperedges
 
 """
-struct Hypergraph{T} <: AbstractMatrix{Union{T, Nothing}}
+struct Hypergraph{T,V,E} <: AbstractMatrix{Union{T, Nothing}}
     v2he::Vector{Dict{Int,T}}
     he2v::Vector{Dict{Int,T}}
-    Hypergraph{T}(n, k) where {T<:Real} =
-        new([Dict{Int,T}() for i in 1:n], [Dict{Int,T}() for i in 1:k])
+    vs::Vector{V}
+    hes::Vector{E}
```

Comment:
Looking at it maybe it would be better to make them `Vector{Union{Nothing, E}` and initialize to `nothing`. In this way we can know when metadata is set and when it is not set. Now you initialize it to garbage, but garbage is indistinguishable from a valid value for immutable types.

Reply:
or we can add two more parameters `v_default` and `e_default` and leave this decision for the user. Then they could write something like this (an we will need to document it):
```julia
h3 = Hypergraph{Float64,Union{String,Nothing}}(2,2; v_default=nothing)
```
The same pattern has been used for Julia `Array`s.

Reply:
Whatever you prefer. For me having `nothing` as the default is a natural choice (this value exactly is meant to say that there is no value).

If you go the way you propose please remember that we have to make a union of `v_default` type and `V`.